File: abc/089/d.py
# 入力
H, W, D = map(int, input().split())
A = [list(map(int, input().split())) for _ in range(H)]
Q = int(input())
L, R = zip(*(map(int, input().split()) for _ in range(Q)))

# 各整数の座標を調べる
p = {A[i][j]: (i, j) for i in range(H) for j in range(W)}

# 各整数に対する累積和を求める
dp = [0 for _ in range(H * W + 1)]
for x in range(D + 1, H * W + 1):
    y = x - D
    dp[x] = dp[y] + abs(p[x][0] - p[y][0]) + abs(p[x][1] - p[y][1])

# 各クエリに対する解を累積和を用いて求める
ans = '\n'.join(str(dp[r] - dp[l]) for l, r in zip(L, R))

# 出力
print(ans)

File: abc/089/d.rs
#![allow(non_snake_case)]
#![allow(unused_variables)]
#![allow(dead_code)]

fn main() {
    let (H, W, D): (usize, usize, usize) = {
        let mut line: String = String::new();
        std::io::stdin().read_line(&mut line).unwrap();
        let mut iter = line.split_whitespace();
        (
            iter.next().unwrap().parse().unwrap(),
            iter.next().unwrap().parse().unwrap(),
            iter.next().unwrap().parse().unwrap()
        )
    };
    let A: Vec<Vec<usize>> = {
        let mut A = vec![];
        for _ in 0..H {
            let mut line: String = String::new();
            std::io::stdin().read_line(&mut line).unwrap();
            A.push(
                line.split_whitespace()
                    .map(|x| x.parse().unwrap())
                    .collect()
            );
        }
        A
    };
    let Q: usize = {
        let mut line: String = String::new();
        std::io::stdin().read_line(&mut line).unwrap();
        line.trim().parse().unwrap()
    };
    let (L, R): (Vec<usize>, Vec<usize>) = {
        let (mut L, mut R) = (vec![], vec![]);
        for _ in 0..Q {
            let mut line: String = String::new();
            std::io::stdin().read_line(&mut line).unwrap();
            let mut iter = line.split_whitespace();
            L.push(iter.next().unwrap().parse().unwrap());
            R.push(iter.next().unwrap().parse().unwrap());
        }
        (L, R)
    };

    let N = H * W;
    let mut X = vec![0; N + 1];
    let mut Y = vec![0; N + 1];
    for i in 0..H {
        for j in 0..W {
            let a = A[i][j];
            X[a] = j as i64;
            Y[a] = i as i64;
        }
    }
    let mut dp = vec![0; N + 1];
    for i in (D + 1)..(N + 1) { dp[i] = dp[i - D] + (X[i] - X[i - D]).abs() + (Y[i] - Y[i - D]).abs(); }

    let ans = (0..Q)
        .map(|i| (dp[R[i]] - dp[L[i]]).to_string())
        .collect::<Vec<_>>()
        .join("\n");

    println!("{}", ans);
}